File: JavaScript/ExampleTests/common/RequestTyped.js
import { RequestString } from "./RequestString.js";
import * as _json from "./json.js";

export class RequestTyped {
  constructor(method = "", page = "", address = "", benchmark = "", format = "") {
    this.method = method;
    this.page = page;
    this.address = address;
    this.benchmark = benchmark;
    this.format = format;
  }

  static fromStringObj(s) {
    return new RequestTyped(
      s.method,
      s.page,
      s.address,
      s.benchmark,
      s.format
    );
  }

  toStringObj() {
    return new RequestString(
      String(this.method),
      String(this.page),
      String(this.address),
      String(this.benchmark),
      String(this.format)
    );
  }

  static toStringList(list) {
    return list.map(t => t.toStringObj());
  }

  static fromStringList(list) {
    return list.map(s => RequestTyped.fromStringObj(s));
  }

  toJsonValue() {
    return {
      method: this.method,
      page: this.page,
      address: this.address,
      benchmark: this.benchmark,
      format: this.format,
    };
  }

  toJSON() { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m) {
    return new RequestTyped(
      _json.asString(_json.require(m, "method"), "method"),
      _json.asString(_json.require(m, "page"), "page"),
      _json.asString(_json.require(m, "address"), "address"),
      _json.asString(_json.require(m, "benchmark"), "benchmark"),
      _json.asString(_json.require(m, "format"), "format")
    );
  }

  static fromJSON(text) { return RequestTyped.fromJsonValue(_json.parse(text)); }

  static toJSONList(items) {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text) {
    const raw = _json.asArray(_json.parse(text), "RequestTyped");
    return raw.map((e) => RequestTyped.fromJsonValue(e));
  }

  toString() {
    return `Method=${this.method}, Page=${this.page}, Address=${this.address}, Benchmark=${this.benchmark}, Format=${this.format}`;
  }

  equals(other) {
    if (!(other instanceof RequestTyped)) return false;
    return this.method === other.method
      && this.page === other.page
      && this.address === other.address
      && this.benchmark === other.benchmark
      && this.format === other.format;
  }
}

File: Rust/ExampleTests/common/request_typed.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::json;
use super::request_string::RequestString;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct RequestTyped {
    pub method: String,
    pub page: String,
    pub address: String,
    pub benchmark: String,
    pub format: String,
}

impl RequestTyped {
    pub fn from_str_struct(s: &RequestString) -> Self {
        Self {
            method: s.method.clone(),
            page: s.page.clone(),
            address: s.address.clone(),
            benchmark: s.benchmark.clone(),
            format: s.format.clone(),
        }
    }

    pub fn to_str_struct(&self) -> RequestString {
        RequestString {
            method: self.method.clone(),
            page: self.page.clone(),
            address: self.address.clone(),
            benchmark: self.benchmark.clone(),
            format: self.format.clone(),
        }
    }

    pub fn to_string_list(list: &[RequestTyped]) -> Vec<RequestString> {
        list.iter().map(|t| t.to_str_struct()).collect()
    }

    pub fn from_string_list(list: &[RequestString]) -> Vec<RequestTyped> {
        list.iter().map(RequestTyped::from_str_struct).collect()
    }

    pub fn to_json_value(&self) -> json::Value {
        json::Value::Object(vec![
            ("Method".to_string(), json::Value::Str(self.method.clone())),
            ("Page".to_string(), json::Value::Str(self.page.clone())),
            ("Address".to_string(), json::Value::Str(self.address.clone())),
            ("Benchmark".to_string(), json::Value::Str(self.benchmark.clone())),
            ("Format".to_string(), json::Value::Str(self.format.clone())),
        ])
    }

    pub fn to_json(&self) -> String {
        json::write(&self.to_json_value())
    }

    pub fn from_json_value(v: &json::Value) -> json::JsonResult<Self> {
        Ok(Self {
            method: json::as_string(json::require(v, "Method")?, "Method")?,
            page: json::as_string(json::require(v, "Page")?, "Page")?,
            address: json::as_string(json::require(v, "Address")?, "Address")?,
            benchmark: json::as_string(json::require(v, "Benchmark")?, "Benchmark")?,
            format: json::as_string(json::require(v, "Format")?, "Format")?,
        })
    }

    pub fn from_json(text: &str) -> json::JsonResult<Self> {
        Self::from_json_value(&json::parse(text)?)
    }

    pub fn to_json_list(list: &[RequestTyped]) -> String {
        json::write(&json::Value::Array(
            list.iter().map(|item| item.to_json_value()).collect(),
        ))
    }

    pub fn from_json_list(text: &str) -> json::JsonResult<Vec<RequestTyped>> {
        let root = json::parse(text)?;
        let items = json::as_array(&root, "RequestTyped")?;
        let mut result = Vec::with_capacity(items.len());
        for e in items {
            result.push(Self::from_json_value(e)?);
        }
        Ok(result)
    }
}
